src/effects/RealtimeEffectStateUI.cpp

```diff
@@ -0,0 +1,160 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  RealtimeEffectStateUI.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#include "RealtimeEffectStateUI.h"
+
+#include <cassert>
+
+#include "EffectUI.h"
+#include "RealtimeEffectState.h"
+
+#include "EffectManager.h"
+#include "ProjectWindow.h"
+#include "Track.h"
+
+namespace
+{
+const RealtimeEffectState::RegisteredFactory realtimeEffectStateUIFactory { [](auto& state)
+   { return std::make_unique<RealtimeEffectStateUI>(state); }
+};
+} // namespace
+
+
+RealtimeEffectStateUI::RealtimeEffectStateUI(RealtimeEffectState& state)
+    : mRealtimeEffectState(state)
+{
+}
+
+RealtimeEffectStateUI::~RealtimeEffectStateUI()
+{
+   Hide();
+}
+
+bool RealtimeEffectStateUI::IsShown() const noexcept
+{
+   return mEffectUIHost != nullptr;
+}
+
+void RealtimeEffectStateUI::Show(AudacityProject& project)
+{
+   if (mEffectUIHost != nullptr && mEffectUIHost->IsShown())
+   {
+      // Bring the host to front
+      mEffectUIHost->Raise();
+      return;
+   }
+
+   const auto ID = mRealtimeEffectState.GetID();
+   const auto effectPlugin = EffectManager::Get().GetEffect(ID);
+
+   if (effectPlugin == nullptr)
+      return;
+
+   EffectUIClientInterface* client = effectPlugin->GetEffectUIClientInterface();
+
+   // Effect has no client interface
+   if (client == nullptr)
+      return;
+
+   // mUIHost will acquire the correct instance
+   // and will manage its lifetime
+   auto& projectWindow = ProjectWindow::Get(project);
+
+   std::shared_ptr<EffectInstance> pInstance;
+
+   auto access = mRealtimeEffectState.GetAccess();
+
+   // EffectUIHost invokes shared_from_this on access
+   Destroy_ptr<EffectUIHost> dlg(safenew EffectUIHost(
+      &projectWindow, project, *effectPlugin, *client, pInstance, *access,
+      mRealtimeEffectState.shared_from_this()));
+
+   if (!dlg->Initialize())
+      return;
+
+   // Dialog is owned by its parent now!
+   mEffectUIHost = dlg.release();
+
+   UpdateTitle();
+
+   client->ShowClientInterface(
+      projectWindow, *mEffectUIHost, mEffectUIHost->GetValidator(), false);
+
+   // The dialog was modal? That shouldn't have happened
+   if (mEffectUIHost == nullptr || !mEffectUIHost->IsShown())
+   {
+      assert(false);
+      mEffectUIHost = {};
+   }
+
+   mProjectWindowDestroyedSubscription = projectWindow.Subscribe(
+      [this](ProjectWindowDestroyedMessage) { Hide(); });
```

Comment:
Did you intend to hide a non-modal dialog but not destroy it?

The closing of the project window does cause the dialog to be destroyed, because it sets its parent pointer to the ProjectWindow.

There are other non-modal dialogs associated with a project -- like Tools > Macros... -- and nothing special is needed to ensure that they are cleaned up.

Reply:
It fixes an assert in `EffectUIHost::~EffectUIHost()`. Is this assert important? You tell me.

Reply:
Okay, and maybe that old assertion is an inappropriate one.  But let it all stand.